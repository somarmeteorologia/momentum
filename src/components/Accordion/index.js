import React, { memo, useState, useContext } from 'react'
import PropTypes from 'prop-types'
import styled, { css, ThemeContext } from 'styled-components'
import { ifProp, theme } from 'styled-tools'
import { motion } from 'framer-motion'

import { Icon } from '@components/Icon'

const Wrapper = styled.div`
  width: ${ifProp('full', '100%', '300px')};

  ${ifProp(
    'auto',
    css`
      min-width: 300px;
      width: auto;
    `
  )}
`

const Container = styled.div`
  font-family: ${theme('font.family.inter')};
  width: 100%;

  &:not(:last-child) {
    ${Headerable} {
      margin-bottom: 5px;
    }
  }

  .body {
    margin-top: 5px;
    padding: 14px 16px;
    font-size: ${theme('font.size.twelve')};
    font-weight: ${theme('font.weight.normal')};
    color: ${theme('accordion.text.primary')};
    background-color: ${theme('accordion.bg.primary')};
    border-radius: ${theme('border.radius.four')};
    line-height: 150%;
  }
`

const Headerable = styled.div`
  height: 40px;
  padding: 14px 48px 14px 16px;
  position: relative;
  display: flex;
  align-items: center;
  font-size: ${theme('font.size.fourteen')};
  font-weight: ${theme('font.weight.bold')};
  color: ${theme('accordion.text.primary')};
  background-color: ${theme('accordion.bg.primary')};
  border-radius: ${theme('border.radius.four')};
  cursor: pointer;

  .icon {
    position: absolute;
    right: 16px;
  }
`

const Bodyable = motion.div

// ({
// })

const Iconable = styled(motion.span)`
  width: 16px;
  height: 16px;
`

export const Accordion = memo(({ accordions, auto, full }) => {
  const { accordion } = useContext(ThemeContext)
  const [list, setList] = useState(accordions)

  const setActive = id => {
    const newList = list.map(item =>
      item.id === id ? { ...item, active: !item.active } : { ...item }
    )

    setList(newList)
  }

  return (
    <Wrapper auto={auto} full={full}>
      {list.map(({ id, active, Header, Body }) => {
        return (
          <Container key={id}>
            <Headerable
              data-testid="header-accordion"
              onClick={() => setActive(id)}
            >
              {Header}

              <Iconable
                className="icon"
                initial="close"
                animate={active ? 'open' : 'close'}
                variants={{
                  open: {
                    rotate: 0
                  },
                  close: {
                    rotate: '45deg'
                  }
                }}
              >
                <Icon
                  name={Icon.name.plus}
                  width={16}
                  height={16}
                  color={
                    active ? accordion.icon.secondary : accordion.icon.primary
                  }
                />
              </Iconable>
            </Headerable>
            <Bodyable
              animate={active ? 'visible' : 'hidden'}
              className="body"
              data-testid="body-accordion"
              variants={{
                visible: {
                  display: 'block',
                  opacity: 1
                },
                hidden: {
                  transitionEnd: {
                    display: 'none'
                  },
                  opacity: 0
                }
              }}
            >
              {Body}
            </Bodyable>
          </Container>
        )
      })}
    </Wrapper>
  )
})

Accordion.defaultProps = {
  auto: false,
  full: false
}

Accordion.propTypes = {
  accordions: PropTypes.array.isRequired,
  auto: PropTypes.bool,
  full: PropTypes.bool
}
