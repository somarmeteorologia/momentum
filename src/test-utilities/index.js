import React from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'

import { Theme } from '@components/Theme'

function renderWithTheme(ui, theme = Theme.light) {
    return {
        ...render(
            <ThemeProvider theme={theme}>
                {ui}
            </ThemeProvider>
        )
    }
}

export {
    renderWithTheme
}