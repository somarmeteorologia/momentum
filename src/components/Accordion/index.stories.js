import React from 'react'
import styled from 'styled-components'
import { storiesOf } from '@storybook/react'

import { GROUPS } from '@helpers/hierarchySeparators'
import { Container } from '@helpers/components/Container'

import { Reset } from '@components/Reset'
import { Accordion } from '@components/Accordion'

const Containerable = styled(Container)`
  flex-wrap: wrap;
`

storiesOf(`${GROUPS.COMPONENTS}|Accordion`, module)
  .addDecorator(story => (
    <Containerable>
      <Reset />
      {story()}
    </Containerable>
  ))
  .add('Default', () => (
    <>
      <Accordion
        accordions={[
          {
            id: 1,
            active: false,
            text: 'foo',
            Header: <div>Lorem ipsum dolor sit amet</div>,
            Body: (
              <div>
                Excepteur sint occaecat cupidatat non proident, sunt in culpa
                qui officia deserunt mollit anim id est laborum.
              </div>
            )
          },
          {
            id: 2,
            active: false,
            text: 'bar',
            Header: <div>Sed ut perspiciatis unde omnis</div>,
            Body: (
              <div>
                Ut enim ad minima veniam, quis nostrum exercitationem ullam
                corporis suscipit laboriosam, nisi ut aliquid ex ea commodi
                consequatur?
              </div>
            )
          }
        ]}
      />
    </>
  ))
  .add('Auto', () => (
    <>
      <Accordion
        accordions={[
          {
            id: 1,
            active: false,
            text: 'foo',
            Header: (
              <div>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit.
                Phasellus bibendum sagittis massa eu ullamcorper
              </div>
            ),
            Body: (
              <div>
                Excepteur sint occaecat cupidatat non proident, sunt in culpa
                qui officia deserunt mollit anim id est laborum.
              </div>
            )
          },
          {
            id: 2,
            active: false,
            text: 'bar',
            Header: <div>Sed ut perspiciatis unde omnis</div>,
            Body: (
              <div>
                Ut enim ad minima veniam, quis nostrum exercitationem ullam
                corporis suscipit laboriosam, nisi ut aliquid ex ea commodi
                consequatur?
              </div>
            )
          }
        ]}
        auto
      />
    </>
  ))
  .add('Full', () => (
    <>
      <Accordion
        accordions={[
          {
            id: 1,
            active: false,
            text: 'foo',
            Header: (
              <div>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit.
                Phasellus bibendum sagittis massa eu ullamcorper
              </div>
            ),
            Body: (
              <div>
                Excepteur sint occaecat cupidatat non proident, sunt in culpa
                qui officia deserunt mollit anim id est laborum.
              </div>
            )
          },
          {
            id: 2,
            active: false,
            text: 'bar',
            Header: <div>Sed ut perspiciatis unde omnis</div>,
            Body: (
              <div>
                Ut enim ad minima veniam, quis nostrum exercitationem ullam
                corporis suscipit laboriosam, nisi ut aliquid ex ea commodi
                consequatur?
              </div>
            )
          }
        ]}
        full
      />
    </>
  ))
