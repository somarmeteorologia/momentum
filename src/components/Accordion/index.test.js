import React from 'react'
import { cleanup, fireEvent, waitForElement } from '@testing-library/react'

import { Accordion } from '@components/Accordion'
import { renderWithTheme } from '../../test-utilities'

const accordions = [
    {
        id: 1,
        active: false,
        text: 'foo',
        Header: <div>Header</div>,
        Body: (
            <div>
                Body.
          </div>
        )
    }
]

afterEach(cleanup)

test('expect render with theme', () => {
    const { container } = renderWithTheme(<Accordion accordions={accordions} />)
    expect(container.firstChild).toMatchSnapshot()
})

test('expect open body with click', async () => {
    const { getByText, container } = renderWithTheme(<Accordion accordions={accordions} />)
    const header = getByText('Header')
    fireEvent.click(header)
    const body = await waitForElement(() => container.getElementsByClassName('body')[0])
    expect(body).toHaveStyle('display : block')
})

test('expect full size', () => {
    const { container } = renderWithTheme(<Accordion full accordions={accordions} />)
    expect(container.firstChild).toHaveStyleRule('width', '100%')
})