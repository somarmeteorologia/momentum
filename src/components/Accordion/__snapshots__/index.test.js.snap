// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect render with theme 1`] = `
.c4 {
  display: inline-block;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 0;
}

.c0 {
  width: 300px;
}

.c1 {
  font-family: Inter,sans-serif;
  width: 100%;
}

.c1:not(:last-child) {
  margin-bottom: 5px;
}

.c1 .body {
  margin-top: 5px;
  padding: 14px 16px;
  font-size: 0.75rem;
  color: #080B19;
  background-color: #f1f3fA;
  border-radius: 0.25rem;
  line-height: 150%;
}

.c2 {
  height: 40px;
  padding: 14px 48px 14px 16px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.875rem;
  font-weight: 600;
  color: #080B19;
  background-color: #f1f3fA;
  border-radius: 0.25rem;
  cursor: pointer;
}

.c2 .icon {
  position: absolute;
  right: 16px;
}

.c3 {
  width: 16px;
  height: 16px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div>
        Header
      </div>
      <span
        class="icon c3"
        style="transform: none;"
      >
        <div
          class="c4"
          height="16"
          width="16"
        >
          <svg
            fill="none"
            height="16"
            viewBox="0 0 15 16"
            width="16"
          >
            <path
              clip-rule="evenodd"
              d="M7.77817 1.63604L6.36396 1.63604L6.36396 7.29289H0.707107V8.70711H6.36396L6.36396 14.364L7.77817 14.364V8.70711L13.435 8.70711V7.29289L7.77817 7.29289V1.63604Z"
              fill="#6479cb"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </span>
    </div>
    <div
      class="body"
      data-testid="body-accordion"
      style="opacity: 0; display: none;"
    >
      <div>
        Body.
      </div>
    </div>
  </div>
</div>
`;
